app/models/mapping.rb

```diff
@@ -73,68 +84,66 @@ def save_and_set_facls
     end
   end
 
-  # Handle FACL setting / removal
-
-  before_save do |mapping|
-    # Short circuit if the admin user cannot set FACLS
-    if can_change_facls?(mapping.dataset)
-      errors = add_user_facls(mapping)
-      unless errors.nil?
-        raise errors
-      end
+  def should_add_facl?(pathname)
+    # Using writable would try setting a FACL on /dev/null...
+    if !pathname.owned?
+      return false
     end
-  end
 
-  before_destroy do |mapping|
-    # Short circuit if the admin user cannot set FACLS
-    if can_change_facls?(mapping.dataset)
-      errors = remove_user_facls(mapping)
-      unless errors.nil?
-        raise errors
-      end
+    # Protected from throwing InvalidPath by the writable? check
+    if rx_facl_exists?(pathname)
+      return false
     end
-  end
-
-  # Update the YAML Dump
 
-  after_create do |mapping|
-    Mapping.dump_to_yaml
+    return true
   end
```

Comment:
This could be simplified without losing readability, and comments added to right or on top:

```ruby
def should_add_facl?(pathname)
  pathname.owned? && ! rx_facl_exists?(pathname) # note: calling owned? before rx_facl_exists? protects rx_facl_exists? from throwing InvalidPath
end
```
This is easier to read at a glance and see what is happening.

I think that while its interesting that `# Using writable would try setting a FACL on /dev/null...`  in this case even if it didn't the logic would still be preferred to only set the facl if you own the file (thus testing for ownership). I might have write access to a client's dataset because his permissions were generously set, but its still his dataset so his responsibility to manage access.

I know some people will go as far as to actually use refinements, which is scoped monkey patching, to add methods these methods on Pathname itself i.e. `Pathname#rx_facl_exists?` which then gives you the ability to write:

```ruby
def should_add_facl?(pathname)
  pathname.owned? && ! pathname.rx_facl_exists?
end
```

But the problem is that those refinements are only available in the place where we explicitly choose to use the refinement. And unless I know the Pathname method list on the back of my hand, I might try to use that cool new Pathname method somewhere else and get a "method missing error" because I didn't realize it was defined in a refinement.

So IMHO what you have done, defining `should_add_facl?` and `rx_facl_exists?` as methods that accept pathname is perfect. 

And you made the right call: isolating the logic under test `should_add_facl?` from the method that has side effects `add_rx_facl` is exactly what you should do.

Reply:
I find the idea of classes being open for modification without having to subclass in Ruby odd, and I don't think that it is something that I'll take up as a practice.

I do quite like the ability to use question marks in method names (or exclamation marks).

Reply:
That classes are open for modification does have some really great use cases, but have obvious caveats. Some benefits:

1. If you stick to a single Ruby extensions library, like ActiveSupport, all of the extra methods added to base Ruby classes make life really easy. You can't mix large extension libraries unless the authors work hard to not introduce conflicts of course.
2. It makes writing mocking libraries and testing frameworks easier.
3. In Open OnDemand, we add paths under /etc/ood/config to the list of paths Rails looks for initializers, which means that at any time any admin could drop an initializer in that monkey patches (and thus changes) any method on any class in an OnDemand app. Which may be a good or really really bad idea; almost any defect can be temporarily addressed using this approach (I need a fix now can't wait till the next OOD release type of thing). Of course that is a step above modifying the code inline of an app to make it work.
4. It is actually how ActiveRecord and Rails plugins work ([osc_machete_rails](https://github.com/OSC/osc_machete_rails/blob/103bb9b37b40684d4745b7198ebab854a5f6121e/lib/osc_machete_rails/workflow.rb#L246-L247)). We also borrowed the Rails approach to constructing the correct Adapter subclass for databases based on a configuration file dynamically to do so for OodCore job adapters: each adapter creates a "build_adapter" method on the Adapter superclass and a single factory will execute this method and pass the rest of the configuration to this method:

https://github.com/OSC/ood_core/blob/d21a1d62a41b1a3e6d05a28cfe1100bb931597f9/lib/ood_core/job/adapters/lsf.rb#L5-L19

https://github.com/OSC/ood_core/blob/d21a1d62a41b1a3e6d05a28cfe1100bb931597f9/lib/ood_core/job/factory.rb#L30-L36

Though IMHO that is overkill - ood_core is not Rails, we don't need that much flexibility. I'll open an issue to simplify that code.